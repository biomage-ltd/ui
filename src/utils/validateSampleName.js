const errorMessages = {
  MIN_1_CHAR: 'Min 1 alphanumeric character',
  INVALID_CHARS: 'Invalid characters',
  NAME_EXISTS: 'A sample with the same name exists',
};

const validateSampleName = (input, sampleNames) => {
  if (input.length < 1 || !input.match(/[a-zA-Z\d]{1,}/gm)) {
    return errorMessages.MIN_1_CHAR;
  }

  if (!input.match(/^[a-zA-Z\s\d-_]{1,}$/gm)) {
    return errorMessages.INVALID_CHARS;
  }

  if (sampleNames instanceof Array
    && sampleNames.includes(input)) {
    return errorMessages.NAME_EXISTS;
  }

  if (sampleNames instanceof Set
    && sampleNames.has(input)) {
    return errorMessages.NAME_EXISTS;
  }

  return true;
};

export default validateSampleName;
