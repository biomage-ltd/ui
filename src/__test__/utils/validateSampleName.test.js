import validateSampleName from '../../utils/validateSampleName';

describe('validateSampleName unit test', () => {
  it('sample names minimal 1 char ', () => {
    const validName1 = '1';
    expect(validateSampleName(validName1)).toEqual(true);

    const validName2 = ' A';
    expect(validateSampleName(validName2)).toEqual(true);

    const inValidName1 = '';
    expect(validateSampleName(inValidName1)).not.toEqual(true);
    expect(validateSampleName(inValidName1)).toMatchSnapshot();

    const inValidName2 = ' ';
    expect(validateSampleName(inValidName2)).toMatchSnapshot();
  });

  it('sample name should not contain invalid characters', () => {
    const invalidName = 'Yummy: Another great sample!';

    expect(validateSampleName(invalidName)).not.toEqual(true);
    expect(validateSampleName(invalidName)).toMatchSnapshot();
  });

  it('sample name should not be the same with existing samples', () => {
    const validName = 'T2';
    const sampleNames = [
      'WT1',
      'WT2',
      'OK',
    ];
    const inValidName = sampleNames[0];

    expect(validateSampleName(validName, sampleNames)).toEqual(true);
    expect(validateSampleName(inValidName, sampleNames)).not.toEqual(false);
    expect(validateSampleName(inValidName)).toMatchSnapshot();
  });
});
