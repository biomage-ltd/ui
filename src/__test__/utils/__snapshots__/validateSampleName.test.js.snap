// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateSampleName unit test sample name should not be the same with existing samples 1`] = `true`;

exports[`validateSampleName unit test sample name should not contain invalid characters 1`] = `"Invalid characters"`;

exports[`validateSampleName unit test sample names minimal 1 char  1`] = `"Minimal 1 alphanumeric character"`;

exports[`validateSampleName unit test sample names minimal 1 char  2`] = `"Minimal 1 alphanumeric character"`;
